lib/tools/apk-utils.js

```diff
@@ -349,6 +355,54 @@ apkUtilsMethods.installFromDevicePath = async function (apkPathOnDevice, opts =
   }
 };
 
+/**
+ * @typedef {Object} CachingOptions
+ * @property {?number} timeout [adbExecTimeout] - The count of milliseconds to wait until the
+ *                                                app is uploaded to the remote location.
+ */
+
+/**
+ * Caches the given APK at a remote location to speed up further APK deployments.
+ *
+ * @param {string} apkPath - Full path to the apk on the local FS
+ * @param {?CachingOptions} options - Caching options
+ * @returns {string} - Full path to the cached apk on the remote file system
+ */
+apkUtilsMethods.cacheApk = async function (apkPath, options = {}) {
```

Comment:
Could we have one e2e test for this. Maybe something that calls `cacheApk` method `REMOTE_CACHE_SIZE + 1` times and verifies that the cache `/data/local/tmp/appium_cache` only contains `REMOTE_CACHE_SIZE` entries. I'm pretty confident this implementation will work, but I would like something to ensure that if we make adjustments to this algorithm, that it will always clean up for itself and that it won't ever grow indefinitely.

Reply:
Also, maybe a test that validates that the exact same file never gets added to the cache more than once. But that when it gets added back, it moves up in "recently used" priority.

Reply:
added unit test